File: Day-20/C++/main.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios_base::sync_with_stdio(false); 
    cin.tie(NULL);

    int target, n = 1000000;
    cin >> target;

    int presents[n];
    for (int i = 1; i < n; i++) {
        int tmp = i - 1;
        while (tmp < n) {
            presents[tmp] += 10 * i;
            tmp += i;
        }
        if (presents[i - 1] >= target) {
            cout << "Part 1: " << i << endl;
            break;
        }
    }

    fill(presents, presents + n, 0);
    for (int i = 1; i < n; i++) {
        int tmp = i - 1;
        for (int j = 0; j < 50 && tmp < n; j++) {
            presents[tmp] += 11 * i;
            tmp += i;
        }
        if (presents[i - 1] >= target) {
            cout << "Part 2: " << i << endl;
            break;
        }
    }

    return 0;
}

File: Day-20/Rust/main.rs
fn get_input() -> String {
    let mut buffer = String::new();
    std::io::stdin().read_line(&mut buffer).expect("Failed");
    buffer
}

fn main() {
    let target = get_input().trim().parse::<usize>().unwrap();
    let mut presents = [0; 1000000];
    for i in 1..presents.len() {
        for j in (i - 1..presents.len()).step_by(i) {
            presents[j] += 10 * i;
        }
        if presents[i - 1] >= target {
            println!("Part 1: {}", i);
            break;
        }
    }

    presents = [0; 1000000];
    let mut tmp;
    for i in 1..presents.len() {
        tmp = i - 1;
        for _ in 0..50 {
            if tmp >= presents.len() {
                break;
            }
            presents[tmp] += 11 * i;
            tmp += i;
        }
        if presents[i - 1] >= target {
            println!("Part 2: {}", i);
            break;
        }
    }
}